pdcupdater/handlers/modules.py

```diff
@@ -0,0 +1,267 @@
+            log.debug("Reading/parsing {}".format(mmd_yaml))
+            mmd = modulemd.ModuleMetadata()
+            mmd.load(mmd_yaml)
+
+            return mmd
+
+    def create_unreleased_variant(self, pdc, body):
+        """Creates an UnreleasedVariant for a module in PDC. Checks out the
+        module metadata from the supplied SCM repository (currently only
+        anonymous GIT is supported)."""
+
+        scmurl = body['scmurl']
+        mmd = self.get_mmd_from_scm(scmurl)
+
+        runtime_deps = []
+        for dep, ver in mmd.requires.items():
+            if ver is not None:
+                runtime_deps.append("{} >= {}".format(dep, ver))
+            else:
+                runtime_deps.append(dep)
+
+        build_deps = []
+        for dep, ver in mmd.buildrequires.items():
+            if ver is not None:
+                build_deps.append("{} >= {}".format(dep, ver))
+            else:
+                build_deps.append(dep)
+
+        koji_tag = body['koji_tag']
+        variant_uid = variant_name = body['module_uid']
+        variant_id = variant_uid.lower()
+        variant_version = body['module_version']
+        variant_release = body['module_release']
+
+        unreleased_variant = pdc['unreleasedvariants']._({
+            'variant_id': variant_id,
+            'variant_uid': variant_uid,
+            'variant_name': variant_name,
+            'variant_version': variant_version,
+            'variant_release': variant_release,
+            'variant_type': 'module',
+            'koji_tag': koji_tag,
+            'runtime_deps': runtime_deps,
+            'build_deps': build_deps,
+        })
+
+        return unreleased_variant
+
+    def get_or_create_unreleased_variant(self, pdc, body):
+        variant_uid = body['module_uid']
+        variant_id = variant_uid.lower()
+
+        try:
+            unreleased_variant = pdc['unreleasedvariants'][variant_id]._()
+        except beanbag.BeanBagException as e:
+            if e.response.status_code != 404:
+                raise
+            # a new module!
+            unreleased_variant = self.create_unreleased_variant(pdc, body)
+        return unreleased_variant
+
+    def handle_new_tree(self, pdc, body, unreleased_variant):
+        topdir = body['topdir']
+        tree_id = os.path.basename(topdir)
+
+        log.debug("Trying to import tree from topdir '{}'".format(topdir))
+
+        m = self.tree_id_re.match(tree_id)
+        if not m:
+            log.error("Unexpected tree id: '{}'".format(tree_id))
+            return
+
+        tree_date = m.group('date')
+        tree_date = (
+            tree_date[0:4] + "-" + tree_date[4:6] + "-" + tree_date[6:8])
+
+        # avoid adding trees twice
+        try:
+            pdc['trees'][tree_id]._()
+        except beanbag.BeanBagException as e:
+            if e.response.status_code != 404:
+                raise
+        else:
+            log.info("Tree exists already, skipping: {}".format(tree_id))
```

Comment:
NB: it doesn't check if the tree referenced in the message and the tree stored in PDC already are the same. Can you imagine a scenario where this might be the case, where we should check it?